jax/lax/lax.py

```diff
@@ -2593,6 +2589,48 @@ def _is_singleton_reshape(old, new):
     else:
       return None
 
+def _reshape_sharded_device_array(array, new_sizes, old_sizes):
+  """Returns None if `array` could not be efficiently reshaped.
+
+  This function is primarily to support soft_pmap, although these optimizations
+  could be useful when directly calling reshape as well.
+  """
+  # TODO(skye): the axis split/merge logic below assumes that
```

Comment:
I'll have more general logic in the replication changes.

Reply:
TODO(jekbradbury) then